README.md
# tokenstream

[![GitHub Actions](https://github.com/vberlier/tokenstream/workflows/CI/badge.svg)](https://github.com/vberlier/tokenstream/actions)
[![PyPI](https://img.shields.io/pypi/v/tokenstream.svg)](https://pypi.org/project/tokenstream/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/tokenstream.svg)](https://pypi.org/project/tokenstream/)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)

> A versatile token stream for handwritten parsers.

```python
from tokenstream import TokenStream

def parse_sexp(stream: TokenStream):
    """A basic S-expression parser."""
    with stream.syntax(brace=r"\(|\)", number=r"\d+", name=r"\w+"):
        brace, number, name = stream.expect(("brace", "("), "number", "name")
        if brace:
            return [parse_sexp(stream) for _ in stream.peek_until(("brace", ")"))]
        elif number:
            return int(number.value)
        elif name:
            return name.value

print(parse_sexp(TokenStream("(hello (world 42))")))  # ['hello', ['world', 42]]
```

## Introduction

Writing recursive-descent parsers by hand can be quite elegant but it's often a bit more verbose than expected, especially when it comes to handling indentation and reporting proper syntax errors. This package provides a powerful general-purpose token stream that addresses these issues and more.

### Features

- Define the set of recognizable tokens dynamically with regular expressions
- Transparently skip over irrelevant tokens
- Expressive API for matching, collecting, peeking, and expecting tokens
- Clean error reporting with line numbers and column numbers
- Contextual support for indentation-based syntax
- Checkpoints for backtracking parsers
- Works well with Python 3.10+ match statements

Check out the [`examples`](https://github.com/vberlier/tokenstream/tree/main/examples) directory for practical examples.

## Installation

The package can be installed with `pip`.

```bash
pip install tokenstream
```

## Getting started

You can define tokens with the `syntax()` method. The keyword arguments associate regular expression patterns to token types. The method returns a context manager during which the specified tokens will be recognized.

```python
stream = TokenStream("hello world")

with stream.syntax(word=r"\w+"):
    print([token.value for token in stream])  # ['hello', 'world']
```

Check out the full [API reference](https://vberlier.github.io/tokenstream/api_reference/) for more details.

### Expecting tokens

The token stream is iterable and will yield all the extracted tokens one after the other. You can also retrieve tokens from the token stream one at a time by using the `expect()` method.

```python
stream = TokenStream("hello world")

with stream.syntax(word=r"\w+"):
    print(stream.expect().value)  # "hello"
    print(stream.expect().value)  # "world"
```

The `expect()` method lets you ensure that the extracted token matches a specified type and will raise an exception otherwise.

```python
stream = TokenStream("hello world")

with stream.syntax(number=r"\d+", word=r"\w+"):
    print(stream.expect("word").value)  # "hello"
    print(stream.expect("number").value)  # UnexpectedToken: Expected number but got word 'world'
```

### Filtering the stream

Newlines and whitespace are ignored by default. You can reject interspersed whitespace by intercepting the built-in `newline` and `whitespace` tokens.

```python
stream = TokenStream("hello world")

with stream.syntax(word=r"\w+"), stream.intercept("newline", "whitespace"):
    print(stream.expect("word").value)  # "hello"
    print(stream.expect("word").value)  # UnexpectedToken: Expected word but got whitespace ' '
```

The opposite of the `intercept()` method is `ignore()`. It allows you to ignore tokens and handle comments pretty easily.

```python
stream = TokenStream(
    """
    # this is a comment
    hello # also a comment
    world
    """
)

with stream.syntax(word=r"\w+", comment=r"#.+$"), stream.ignore("comment"):
    print([token.value for token in stream])  # ['hello', 'world']
```

### Indentation

To enable indentation you can use the `indent()` method. The stream will now yield balanced pairs of `indent` and `dedent` tokens when the indentation changes.

```python
source = """
hello
    world
"""
stream = TokenStream(source)

with stream.syntax(word=r"\w+"), stream.indent():
    stream.expect("word")
    stream.expect("indent")
    stream.expect("word")
    stream.expect("dedent")
```

To prevent some tokens from triggering unwanted indentation changes you can use the `skip` argument.

```python
source = """
hello
        # some comment
    world
"""
stream = TokenStream(source)

with stream.syntax(word=r"\w+", comment=r"#.+$"), stream.indent(skip=["comment"]):
    stream.expect("word")
    stream.expect("comment")
    stream.expect("indent")
    stream.expect("word")
    stream.expect("dedent")
```

### Checkpoints

The `checkpoint()` method returns a context manager that resets the stream to the current token at the end of the `with` statement. You can use the returned `commit()` function to keep the state of the stream at the end of the `with` statement.

```python
stream = TokenStream("hello world")

with stream.syntax(word=r"\w+"):
    with stream.checkpoint():
        print([token.value for token in stream])  # ['hello', 'world']
    with stream.checkpoint() as commit:
        print([token.value for token in stream])  # ['hello', 'world']
        commit()
    print([token.value for token in stream])  # []
```

### Match statements

Match statements make it very intuitive to process tokens extracted from the token stream. If you're using Python 3.10+ give it a try and see if you like it.

```python
from tokenstream import TokenStream, Token

def parse_sexp(stream: TokenStream):
    """A basic S-expression parser that uses Python 3.10+ match statements."""
    with stream.syntax(brace=r"\(|\)", number=r"\d+", name=r"\w+"):
        match stream.expect_any(("brace", "("), "number", "name"):
            case Token(type="brace"):
                return [parse_sexp(stream) for _ in stream.peek_until(("brace", ")"))]
            case Token(type="number") as number :
                return int(number.value)
            case Token(type="name") as name:
                return name.value
```

## Contributing

Contributions are welcome. Make sure to first open an issue discussing the problem or the new feature before creating a pull request. The project uses [`poetry`](https://python-poetry.org/).

```bash
$ poetry install
```

You can run the tests with `poetry run pytest`.

```bash
$ poetry run pytest
```

The project must type-check with [`pyright`](https://github.com/microsoft/pyright). If you're using VSCode the [`pylance`](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance) extension should report diagnostics automatically. You can also install the type-checker locally with `npm install` and run it from the command-line.

```bash
$ npm run watch
$ npm run check
$ npm run verifytypes
```

The code follows the [`black`](https://github.com/psf/black) code style. Import statements are sorted with [`isort`](https://pycqa.github.io/isort/).

```bash
$ poetry run isort tokenstream examples tests
$ poetry run black tokenstream examples tests
$ poetry run black --check tokenstream examples tests
```

---

License - [MIT](https://github.com/vberlier/tokenstream/blob/main/LICENSE)


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.7.0 (2023-08-03)
### Feature
* Add preprocessor ([`2a5573c`](https://github.com/vberlier/tokenstream/commit/2a5573c33f9bf90986f12f30c7e903743ee40791))

## v1.6.0 (2023-08-02)
### Feature
* Add token_handler ([`a6b1425`](https://github.com/vberlier/tokenstream/commit/a6b14252af5fdf3c4eb66aa980aa2446822cb702))

## v1.5.0 (2023-01-05)
### Feature
* Make it possible to swap the underlying regex module ([`b47995e`](https://github.com/vberlier/tokenstream/commit/b47995ee40c0e1c43afa68bc891770670e0e1502))

### Fix
* Allow expect_eof() to succeed even if the "eof" token was already consumed ([`7a8e799`](https://github.com/vberlier/tokenstream/commit/7a8e7995e3e9d74a7146e15c30440795ce026fcc))
* Update typing ([`0ce06c7`](https://github.com/vberlier/tokenstream/commit/0ce06c7923a694496382060646d91317f6b8bed9))

## v1.4.2 (2022-06-02)
### Fix
* Update mudkip ([`a58854c`](https://github.com/vberlier/tokenstream/commit/a58854cb7c3a699653d02dc4de82d21d45f637f9))

## v1.4.1 (2022-05-02)
### Fix
* Take into account repeating newlines before emitting dedent tokens ([`109552c`](https://github.com/vberlier/tokenstream/commit/109552c2b363359577f9afa1071d347e8ab55499))

## v1.4.0 (2022-04-12)
### Feature
* Add notes ([`8be67ce`](https://github.com/vberlier/tokenstream/commit/8be67ce6c4c942689fe93ee73d83a2eed19030ae))

## v1.3.4 (2022-04-06)
### Fix
* Better mechanism for tracking alternative errors ([`4e0c3f2`](https://github.com/vberlier/tokenstream/commit/4e0c3f2e5a72e095c57a84ba5c59414151b06b19))

## v1.3.3 (2022-04-02)
### Fix
* Keep track of alternative errors ([`9bee458`](https://github.com/vberlier/tokenstream/commit/9bee458408166badad4795e8174935ea615ea055))

## v1.3.2 (2022-02-19)
### Fix
* Update docs ([`a788ff1`](https://github.com/vberlier/tokenstream/commit/a788ff194eb630555a01ba9837a160e40f97dbec))

## v1.3.1 (2021-12-06)
### Fix
* Sort expected patterns in explanation ([`c1c26cb`](https://github.com/vberlier/tokenstream/commit/c1c26cb874229a5e0319c882f3b6e5567fd14870))

## v1.3.0 (2021-12-01)
### Feature
* Make it possible to copy the stream ([`5a4a41c`](https://github.com/vberlier/tokenstream/commit/5a4a41c8f4a137a9c98bdda856cf4a4ba2169d0b))

## v1.2.7 (2021-11-30)
### Fix
* Get() method was returning a list instead of a single token ([`accf784`](https://github.com/vberlier/tokenstream/commit/accf7843f1263dbba6aa0893a157be608ed1758a))

## v1.2.6 (2021-11-25)
### Fix
* Prevent duplicating expected patterns ([`10cec04`](https://github.com/vberlier/tokenstream/commit/10cec045e3658333cf9a381fd166bcb8cf1a3064))

## v1.2.5 (2021-11-25)
### Fix
* Simplify expected_patterns merging ([`ad10e50`](https://github.com/vberlier/tokenstream/commit/ad10e50804770a11c530a8234aeadd6b9561fb0f))

## v1.2.4 (2021-11-25)
### Fix
* Properly prioritize errors that occur later in the stream ([`a665924`](https://github.com/vberlier/tokenstream/commit/a665924a68e43a6f0f7b1b8cc069761b32436323))

## v1.2.3 (2021-10-20)
### Fix
* Set invalid syntax location for unexpected token ([`4231809`](https://github.com/vberlier/tokenstream/commit/423180921e036bdfb0ea3c1025841248aedbb010))

## v1.2.2 (2021-10-09)
### Fix
* Handle windows line endings ([`068144d`](https://github.com/vberlier/tokenstream/commit/068144d0cadad0265ae292ec5dd9a925265735ef))

## v1.2.1 (2021-09-24)
### Fix
* With_horizontal_offset doesn't modify unkown location anymore ([`e17f897`](https://github.com/vberlier/tokenstream/commit/e17f897fec2feb53b0df3c7c62efdb0fe6c1fcd1))

## v1.2.0 (2021-09-19)
### Feature
* Make it possible to know if a checkpoint was rolled back ([`0765e96`](https://github.com/vberlier/tokenstream/commit/0765e96aec93bcc6b0a60956ac509afd4f63e1c6))

## v1.1.0 (2021-09-16)
### Feature
* Add stream.reset() ([`1af0d47`](https://github.com/vberlier/tokenstream/commit/1af0d47b93082ae374ca8edef8f03f8a549e1fdb))

### Fix
* Do not register the first alternative exception twice ([`1bb7102`](https://github.com/vberlier/tokenstream/commit/1bb71027bdf126ae5c439c149b34a99eeb5973fd))

## v1.0.5 (2021-09-15)
### Fix
* Make it possible to use None to disable previous rules ([`66e468a`](https://github.com/vberlier/tokenstream/commit/66e468ad068a3b1c9c34f0e72b63df38b7710a81))

## v1.0.4 (2021-09-14)
### Fix
* Add period to all error messages ([`c1c32b0`](https://github.com/vberlier/tokenstream/commit/c1c32b02732b66ac4080585f1360bfc429c7f0e2))

## v1.0.3 (2021-09-14)
### Fix
* Don't show token value if it's empty ([`6337bfd`](https://github.com/vberlier/tokenstream/commit/6337bfd35f0f51996cfae15174fe4f1b1428b7f2))

## v1.0.2 (2021-09-14)
### Fix
* Don't detect indent when the line is blank ([`175e02e`](https://github.com/vberlier/tokenstream/commit/175e02eadb75d027199f268edbede774632ab34c))
* Properly handle full dedent ([`04b9dab`](https://github.com/vberlier/tokenstream/commit/04b9dab09bf32cb368a60a009650689e7cf24145))
* Add eof and invalid tokens ([`264a398`](https://github.com/vberlier/tokenstream/commit/264a398ad4e55e89e0bfb503db2af98b38492d00))

## v1.0.1 (2021-09-08)
### Fix
* Handle non SourceLocation end_location in set_location ([`f16b378`](https://github.com/vberlier/tokenstream/commit/f16b378fbdc52942bcf14a8d4b41d34158d867d7))

## v1.0.0 (2021-09-08)
### Feature
* Move SourceLocation into its own module ([`95aabae`](https://github.com/vberlier/tokenstream/commit/95aabae6058bb3c13bbd4f15a372d480f94a22b2))

### Breaking
* remove InvalidSyntax.set_location in favor of the general-purpose setlocation function  ([`95aabae`](https://github.com/vberlier/tokenstream/commit/95aabae6058bb3c13bbd4f15a372d480f94a22b2))

## v0.7.5 (2021-09-04)
### Fix
* Combine unexpected token patterns when using the choose() method ([`b8665e8`](https://github.com/vberlier/tokenstream/commit/b8665e8b10e451467e2dae1659e9c999213aa348))

## v0.7.4 (2021-08-30)
### Fix

tokenstream/location.py
__all__ = [
    "SourceLocation",
    "set_location",
    "INITIAL_LOCATION",
    "UNKNOWN_LOCATION",
]


from bisect import bisect
from dataclasses import FrozenInstanceError, replace
from typing import Any, NamedTuple, Sequence, TypeVar

T = TypeVar("T")


class SourceLocation(NamedTuple):
    """Class representing a location within an input string."""

    pos: int
    lineno: int
    colno: int

    @property
    def unknown(self) -> bool:
        """Whether the location is unknown.

        >>> location = UNKNOWN_LOCATION
        >>> location.unknown
        True
        """
        return self.pos < 0

    def format(self, filename: str, message: str) -> str:
        """Return a message formatted with the given filename and the current location.

        >>> SourceLocation(42, 3, 12).format("path/to/file.txt", "Some error message")
        'path/to/file.txt:3:12: Some error message'
        """
        return f"{filename}:{self.lineno}:{self.colno}: {message}"

    def with_horizontal_offset(self, offset: int) -> "SourceLocation":
        """Create a modified source location along the horizontal axis.

        >>> INITIAL_LOCATION.with_horizontal_offset(41)
        SourceLocation(pos=41, lineno=1, colno=42)
        """
        if self.unknown:
            return self
        return SourceLocation(self.pos + offset, self.lineno, self.colno + offset)

    def skip_over(self, value: str) -> "SourceLocation":
        """Return the source location after skipping over a piece of text.

        >>> INITIAL_LOCATION.skip_over("hello\\nworld")
        SourceLocation(pos=11, lineno=2, colno=6)
        """
        return SourceLocation(
            self.pos + len(value),
            self.lineno + value.count("\n"),
            self.colno + len(value)
            if (line_start := value.rfind("\n")) == -1
            else len(value) - line_start,
        )

    def map(
        self,
        input_mappings: Sequence["SourceLocation"],
        output_mappings: Sequence["SourceLocation"],
    ) -> "SourceLocation":
        """Map a source location.

        The mappings must contain corresponding source locations in order.

        >>> INITIAL_LOCATION.map([], [])
        SourceLocation(pos=0, lineno=1, colno=1)
        >>> mappings1 = [SourceLocation(16, 2, 27), SourceLocation(19, 2, 30)]
        >>> mappings2 = [SourceLocation(24, 3, 8), SourceLocation(67, 4, 12)]
        >>> INITIAL_LOCATION.map(mappings1, mappings2)
        SourceLocation(pos=0, lineno=1, colno=1)
        >>> SourceLocation(15, 2, 26).map(mappings1, mappings2)
        SourceLocation(pos=15, lineno=2, colno=26)
        >>> SourceLocation(16, 2, 27).map(mappings1, mappings2)
        SourceLocation(pos=24, lineno=3, colno=8)
        >>> SourceLocation(18, 2, 29).map(mappings1, mappings2)
        SourceLocation(pos=26, lineno=3, colno=10)
        >>> SourceLocation(19, 2, 30).map(mappings1, mappings2)
        SourceLocation(pos=67, lineno=4, colno=12)
        >>> SourceLocation(31, 3, 6).map(mappings1, mappings2)
        SourceLocation(pos=79, lineno=5, colno=6)
        """
        index = bisect(input_mappings, self) - 1
        if index < 0:
            return self
        return self.relocate(input_mappings[index], output_mappings[index])

    def relocate(
        self,
        base_location: "SourceLocation",
        target_location: "SourceLocation",
    ) -> "SourceLocation":
        """Return the current location transformed relative to the target location."""
        pos, lineno, colno = self

        pos = target_location.pos + (pos - base_location.pos)
        lineno = target_location.lineno + (lineno - base_location.lineno)

        if lineno == target_location.lineno:
            colno = target_location.colno + (colno - base_location.colno)

        return SourceLocation(pos, lineno, colno)


INITIAL_LOCATION = SourceLocation(pos=0, lineno=1, colno=1)
UNKNOWN_LOCATION = SourceLocation(pos=-1, lineno=0, colno=0)


def set_location(
    obj: T,
    location: Any = UNKNOWN_LOCATION,
    end_location: Any = UNKNOWN_LOCATION,
) -> T:
    """Set the location and end_location attributes.

    The function returns the given object or a new instance if the object
    is a namedtuple or a frozen dataclass. The location can be copied from another
    object with location and end_location attributes.

    >>> token = Token("number", "123", UNKNOWN_LOCATION, UNKNOWN_LOCATION)
    >>> updated_token = set_location(token, SourceLocation(15, 6, 1))
    >>> updated_token
    Token(type='number', value='123', location=SourceLocation(pos=15, lineno=6, colno=1), end_location=SourceLocation(pos=15, lineno=6, colno=1))
    >>> updated_token = set_location(
    ...     updated_token,
    ...     end_location=updated_token.location.with_horizontal_offset(len(updated_token.value)),
    ... )
    >>> set_location(token, updated_token)
    Token(type='number', value='123', location=SourceLocation(pos=15, lineno=6, colno=1), end_location=SourceLocation(pos=18, lineno=6, colno=4))
    """
    if not isinstance(end_location, SourceLocation):
        end_location = getattr(end_location, "end_location", UNKNOWN_LOCATION)

    if not isinstance(location, SourceLocation):
        if end_location.unknown:
            end_location = getattr(location, "end_location", UNKNOWN_LOCATION)
        location = getattr(location, "location", UNKNOWN_LOCATION)

    if location.unknown:
        location = getattr(obj, "location", location)
    if end_location.unknown:
        end_location = getattr(obj, "end_location", end_location)

    end_location = max(location, end_location)

    if isinstance(obj, tuple):
        return obj._replace(location=location, end_location=end_location)  # type: ignore

    try:
        obj.location = location  # type: ignore
        obj.end_location = end_location  # type: ignore
    except FrozenInstanceError:
        return replace(obj, location=location, end_location=end_location)

    return obj
